### PR TITLE
fix: incorrect region access with split

### PR DESCRIPTION
**Pull request by haraldh**
For `unmap`: instead of inheriting the access of the previous record, the split records gained the default/empty access.
Also check `unmap` with `Access::READ`. This would have failed without the `unmap` fix.

And some minor patches:

* Make Access generic, so users of the crate can define their
    own Access flags.
* fix: add SPDX-License-Identifier  
* fix debug output
* add mixed access ledger test
* add comments and early returns

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -136,23 +103,19 @@ impl<const N: usize> Ledger<N> {
 
     /// Create a new instance. Make sure that `region.start < region.end`,
     /// because otherwise the ledger will exhibit undefined behaviour.
```
Comment (bstrie) [inline src/lib.rs]: I'm concerned at this comment. If the caller of this function needs to uphold that region.start < region.end or else UB will result, then should this be an `unsafe fn`?

File: src/lib.rs [hunk 2]
```diff
@@ -394,34 +357,85 @@ mod tests {
...
+            /// Access::READ access
+            const READ = 1 << 0;
+
+            /// Access::WRITE access
+            const WRITE = 1 << 1;
+
+            /// Execute access
+            const EXECUTE = 1 << 2;
+
+            const MMAPED = 1 << 3;
```
Comment (bstrie) [inline src/lib.rs]: What is this? Should it be spelled `MMAPPED`?

File: src/lib.rs [hunk 3]
```diff
@@ -394,34 +357,85 @@ mod tests {
...
+                if self.contains(Access::WRITE) {
+                    'w'
+                } else {
+                    '-'
+                },
+                if self.contains(Access::EXECUTE) {
+                    'x'
+                } else {
+                    '-'
+                }
```
Comment (bstrie) [inline src/lib.rs:435]: What about the new MMAPED flag?

File: src/lib.rs [hunk 4]
```diff
@@ -667,6 +691,47 @@ mod tests {
...
+            .collect::<Vec<_>>();
+
+        let expected = expected
+            .iter()
+            .cloned()
+            .map(|record| Record {
+                region: Region::new(Address::new(record.0 << 12), Address::new(record.1 << 12)),
+                access: record.2,
+            })
+            .collect::<Vec<_>>();
```
Comment (bstrie) [inline src/lib.rs]: The two statements above look like boilerplate shared among all the tests, do you think they should be split out into `to_regions` and `to_records` functions?

File: src/lib.rs [hunk 5]
```diff
@@ -667,6 +691,47 @@ mod tests {
...
+
+        let mut ledger = MIXED_LEDGER.clone();
+        assert_eq!(ledger.records(), &[LOWER_HALF_R, UPPER_HALF_W]);
```
Comment (bstrie) [inline src/lib.rs]: Is this assert necessary?

File: src/lib.rs [hunk 6]
```diff
@@ -314,20 +314,29 @@ impl<T: LedgerAccess, const N: usize> Ledger<T, N> {
             let record_start = self.records[index].region.start;
             let record_end = self.records[index].region.end;
 
-            if region.end < record_start || region.start > record_end {
+            if region.start >= record_end {
                 // Skip:
+                // [   ]   XXXXX
```
Comment (bstrie) [inline src/lib.rs]: These comments are cute, but they don't explain why some of these cases cause `continue` while others cause `return` (which is what I find confusing). In general I find the control flow in this function confusing, can some of these ifs/continues be condensed into if-else?

**Review by jarkkojs — DISMISSED** [review]

File: src/lib.rs (hunk 5 above)
Comment (haraldh) [reply src/lib.rs]: not really.. was just a blind copy&paste of the other test

File: src/lib.rs (hunk 6 above)
Comment (haraldh) [reply src/lib.rs]: added more

File: src/lib.rs (hunk 4 above)
Comment (haraldh) [reply src/lib.rs]: done

File: src/lib.rs (hunk 3 above)
Comment (haraldh) [reply src/lib.rs:435]: removed

File: src/lib.rs (hunk 2 above)
Comment (haraldh) [reply src/lib.rs]: removed

File: src/lib.rs (hunk 1 above)
Comment (haraldh) [reply src/lib.rs]: it's not `unsafe` and UB in the rust sense... but I fixed it.

File: src/lib.rs (hunk 5 above)
Comment (haraldh) [reply src/lib.rs]: removed

File: src/lib.rs (hunk 6 above)
Comment (haraldh) [reply src/lib.rs]: and change the control flow to a match statement, so we can catch the `unreachable!()` on logic error.

**Review by bstrie — APPROVED** [review]

**Pull request merged**